File: Assignment5_Sangines/CNNByAMParallel2/CNNByAM/FeatureMap.cs
using CNNAM;
using MatrixLibByAM;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CNNByAM
{
    [Serializable]
    class FeatureMap
    {  // featuremap for layer2 in cnn and onwards
        public FeatureMap(int inputDataSize, PoolingType poolingType, ActivationType activationType, int batchSize)
        {
            this.inputDataSize = inputDataSize;
            this.poolingType = poolingType;
            this.activationType = activationType;
            this.batchSize = batchSize;
            DeltaSS = new Matrix[batchSize];
            DeltaCV = new Matrix[batchSize];
            OutPutSS = new Matrix[batchSize];
            ActCV = new Matrix[batchSize];
            APrime = new Matrix[batchSize];
            Sum = new Matrix[batchSize];
        }
        Random rand = new Random((int)(DateTime.Now.Ticks));  // 5000 = seed for repeatable results
        PoolingType poolingType;
        public ActivationType activationType;
        int inputDataSize;
        int batchSize = 0;
        public Matrix[] DeltaSS { get; set; }   // subsampling deltas
        public Matrix[] DeltaCV { get; set; }   // layer deltas
        public Matrix[] OutPutSS { get; set; }  // subsampling layer output
        public Matrix[] ActCV { get; set; }  // Activation function outputs 
        public Matrix[] APrime { get; set; }  // Aprime 
        public Matrix[] Sum { get; set; }  // result after convol, then +  bias
        public double Bias { get; set; } // one bias for the feature map
        public double BiasGrad { get; set; } // one bias grad for the feature map

        public Matrix Evaluate(Matrix inputData, int batchIndex)
        {
            int c2Size = inputData.Rows;
            Matrix Res = new Matrix(c2Size, c2Size);
            Sum[batchIndex] = inputData.Add(Bias);
            if (activationType == ActivationType.SIGMOID)
            {
                ActCV[batchIndex] = Sum[batchIndex].Sigmoid();
                APrime[batchIndex] = 
                    ActCV[batchIndex].ElementByElementMul(ActCV[batchIndex].Mul(-1).Add(1));
            }
            if (activationType == ActivationType.RELU)
            {
                ActCV[batchIndex] = Sum[batchIndex].RELU(); // no aprime for relu, delta is made zero for neg sums
            }
            if (poolingType == PoolingType.AVGPOOLING)
                Res = ActCV[batchIndex].AvgPool();
            OutPutSS[batchIndex] = Res;
            return Res;
        }

    }
}

File: Assignment5_Sangines/CNN_Sangines/CNN_Sangines/FeatureMap.py
import numpy as np
import enumerations

class FeatureMap(object):
    """description of class"""
    def __init__(self, inputDataSize, poolingType, activationFuncType, batchSize=0):
        self.inputDataSize = inputDataSize
        self.poolingType = poolingType
        self.activationFuncType = activationFuncType
        self.batchSize = batchSize
        #---------All of the following will be List of np arrays---------
        self.deltaSS = [batchSize] #Delta at subsampling (pooling)
        self.deltaCV = [batchSize] ##Delta of the layer
        self.sumation = [batchSize] #result after adding biasses
        self.outputSS = [batchSize] #Output after subsampling (pooling)
        self.actDeriv = [batchSize] #Derivative of Activation function
        self.actFuncOut = [batchSize] #Activation Function in current Layer
        #------------------------------------------------------------------
        self.bias = 0.0
        self.biasGrad = 0.0

    def Evaluate(self, inputData, batchIndex):
        self.sumation[batchSize] = inputData + self.bias #Adding bias to each feature map (sumation is a list of np arrays since inputData will be a numpy array)

        if self.activationFuncType == enumerations.ActivationType.SIGMOID:
            self.actFuncOut[batchIndex] = 1/(1 + np.exp(-self.sumation[batchIndex]))
            self.actDeriv[batchIndex] = self.actFuncOut[batchIndex] * (1 - self.actFuncOut[batchIndex])

        elif self.activationFuncType == enumerations.ActivationType.RELU:
            self.actFuncOut[batchIndex] = np.maximum(0, self.sumation[batchIndex])

        if self.poolingType == enumerations.PoolingType.AVGPOOLING:
            res = np.zeros((len(self.actFuncOut[batchIndex])//2,len(self.actFuncOut[batchIndex,0])//2))
            for i in range(len(self.actFuncOut[batchIndex])//2):
                for j in range(len(self.actFuncOut[batchIndex,0])//2):
                    res[i,j] = np.average(self.actDeriv[batchIndex, i*2:i*2+2, j*2:j*2+2])
            self.outputSS[batchIndex] = res
        return 



        


